### ISSUE TITLE
Create proper model for project content.

### ISSUE THREAD

**Issue by weisJ** [MyLibreLab/MyLibreLab#71]
We need to analyse how the current model works.
Currently the elements and the editor are (though not directly) tightly mangled. This is very undesirable when working with the project model on an abstract level (e.g. during simulation).

I haven't yet looked far into the current structure but I think we could work with something like this:
- Each "element" is a node in a graph.
- The nodes have controllers attached (those contain the logic for each node).
- The nodes have some model object attached which describes the node (i.e. name, description, icon, # of pins etc.)

This is just a rough a outline and I am myself not fully satisfied with the approach as the node still holds too much information about the visual aspect (name, description and icon).

Feedback and thoughts appreciated.

**Comment by marko-radosavljevic** [comment]
There is a lot of data/behaviour that would be useful to us, so we can do separation of concerns however we want. In the end we will have many representations of the elements, and that is fine in my opinion.

One of them can be visual/view-like:
- name
- description
- position
- size
- scale
- rotation
- shading
- outline (when selected, for example)
    - color
    - stroke
    - opacity
    - dotted or not
- other things required for drawing
- icons
- pins/connections and their visual data
- who knows what else we might want

These things have no importance for the actual logic of the application, they are just visual representation of the model in the editor.

If we have some complex drawing, we can have drawing/painter representation. If we are dealing with a lot of state dependent behavior, and we most likely will, we can have a state pattern/finite automata for our elements.

**Comment by weisJ** [comment]
The approach I took for the actions was to attach a `Presentation` object to the actions (in particular an `ActionPresentation` object). Each time the action is updated/is displayed an `onUpdate` method in which the actions gets a chance to update its presentation.

The same could be done for the elements. This way the data model isn’t cluttered with information about how it should be displayed (i.e the data model is a separate object and an element consists of a presentation and a data object).

**Comment by marko-radosavljevic** [comment]
Yup, that sounds sane.  Some separation of concerns that resembles things like mvc/mvp/mvvm. `onUpdate` sounds like an observer pattern, etc, classic stuff.

Because there will be a lot of view-like data/behaviour in the future, much more than I mentioned before. It would be a mistake to start with adding one by one to the model. So yeah, I think we agree on the general idea.
